openwisp_monitoring/device/apps.py

```diff
@@ -73,3 +77,63 @@ def trigger_device_recovery_checks(cls, instance, **kwargs):
 
         if cache.get(get_device_recovery_cache_key(device=instance)):
             trigger_device_recovery.delay(pk=instance.pk)
+
+    @classmethod
+    def connect_is_working_changed(self):
+        from openwisp_controller.connection.models import DeviceConnection
+        from openwisp_controller.connection.signals import is_working_changed
+
+        is_working_changed.connect(
+            self.is_working_changed_receiver,
+            sender=DeviceConnection,
+            dispatch_uid='change_health_status',
+        )
+
+    @classmethod
+    def is_working_changed_receiver(cls, instance, is_working, **kwargs):
+        from ..check.models import Check
+        from .tasks import trigger_device_recovery
+
+        device = instance.device
+        device_monitoring = device.monitoring
+        initial_status = device_monitoring.status
+        status = 'ok' if is_working else 'problem'
+        if not is_working:
+            device_monitoring.update_status(status)
+            # Create a related notification explaining why it's not working
+            desc = instance.failure_reason
+            opts = dict(
+                actor=device, level='warning', verb='not working', description=desc
+            )
+            if initial_status == 'ok':
+                trigger_device_recovery.delay(pk=device.pk)
+        else:
+            # create a notification that device is working
+            opts = dict(actor=device, level='info', verb='connected successfully')
+            try:
+                # if checks exist trigger them else, set status as 'ok'
+                Check.objects.get(object_id=instance.device.pk)
+                trigger_device_recovery.delay(pk=device.pk)
+            except Check.DoesNotExist:
+                device_monitoring.update_status(status)
+        opts['data'] = {'email_subject': f'[{status.upper()}] {device}'}
+        cls._notify_users(opts)
+
+    @classmethod
+    def _notify_users(cls, opts):
```

Comment:
this code looks duplicated from `Metric._notify_users`. **Do not duplicate code**.
Create a function which handles both cases and call that in these two places.
Duplicating code is the best way to maintenance hell. Refrain yourself from doing it.

Reply:
I have tried to avoid duplication for most of the code except for creating the `opts`, it was deferring slightly for both though it has a workaround I have left these lines in the function `_notify_users` also to maintain backward compatibality for tests in the `notifications` app